{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5873175-9a34-413a-aac7-48384ed817ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from sklearn.cluster import KMeans\n",
    "import anndata as ad\n",
    "from scipy import sparse\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff2dc88-9b6c-4375-a2af-e6e0320688cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input files \n",
    "region = \"regionJ\"\n",
    "# processed expression matrix split by regions \n",
    "count_file = pd.read_csv(region+\"_sct_240211.csv\", header = 0, index_col = 0)\n",
    "cop = count_file.T\n",
    "# cell type ident for each cell \n",
    "celltype = pd.read_csv(region+\"_celltypes_240418.csv\")\n",
    "# X, Y, Z coordinate of each cell \n",
    "meta = pd.read_csv(region+\"_metadata_240418.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2fc273-d842-4bf4-ad38-a02d962d97c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at flu genes -- normalized values for convex hull plotting only, otherwise use expression values \n",
    "flu = ['NP.gRNA', 'NS1.gRNA']\n",
    "flu_exp = cop[flu].sum(axis=1)\n",
    "flu_norm = flu_exp / np.max(flu_exp)\n",
    "np_level = np.linspace(0, 1, 10)\n",
    "np_thresh_level = np.digitize(flu_norm, np_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd45cbae-c8d0-495f-af3e-bd7a35e9a9c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at Ifn genes -- normalized values for convex hull plotting only, otherwise use expression values \n",
    "ifn = ['Ifnb1', 'Ifnl2', 'Ifna4']\n",
    "ifn_exp = cop[ifn].sum(axis=1)\n",
    "\n",
    "ifn_norm = ifn_exp / np.max(ifn_exp)\n",
    "ifn_level = np.linspace(0, 1, 10)\n",
    "ifn_thresh_level = np.digitize(ifn_norm, ifn_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98d0a2fb-a684-4167-8a3f-26625c5e9c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(flu_exp, ifn_exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afe2d53e-f43d-48cd-903c-078476f71aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create adata object for convex hull plotting \n",
    "from scipy import sparse\n",
    "mtx_count = sparse.csr_matrix(cop) \n",
    "adata = ad.AnnData(mtx_count)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a825bcfb-a914-4e75-8eb4-e79d93be8e0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.index = meta.index\n",
    "adata.obs['region'] = meta['region'].tolist()\n",
    "adata.obs['column'] = meta['x'].tolist()\n",
    "adata.obs['row'] = meta['y'].tolist()\n",
    "adata.obs['volume'] = meta['volume'].tolist()\n",
    "adata.obs['celltype'] = celltype['x'].tolist()\n",
    "adata.obs['nCount_RNA'] = meta['nCount_RNA'].tolist()\n",
    "adata.obs['flu'] = np_thresh_level\n",
    "adata.obs['ifn'] = ifn_thresh_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d80820ed-91cf-4dd9-b6be-cabdfebb250a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createConvexHulls(adata, readspath, scale_t = 1,reads_min = 10, reads_max = 100000):\n",
    "    '''\n",
    "    Create convex hull borders for cells containing between `reads_min` and `reads_max` reads.\n",
    "    Returns hulls and barcodes for the corresponding cells.\n",
    "    @param adata: AnnData object containing spatial data. The `adata.obs.index` should has the $anything-$cell_barcode format.\n",
    "    @param readspath: path to the reads file\n",
    "    @param scale_t: scaling factor for the coordinates\n",
    "    @param reads_min: minimum number of reads for a cell to be included in the convex hull\n",
    "    @param reads_max: maximum number of reads for a cell to be included in the convex hull\n",
    "    e.g.\n",
    "        hulls, barcodes = createConvexHulls(sdata_t,reads_path,scale_t = scale_dict['63'])\n",
    "        sdata_t2 = sdata_t[np.isin(sdata_t.obs.index.values,barcodes),:].copy()\n",
    "        plotHulls(sdata_t2, hulls, barcodes, col_t, ctype_color_dict, figsize_scale = 1/200,output_path_t=f'{output_path_t}/{sample_t}_{tech_t}_convex_hull.pdf',scale_bar_t=[200,'200 Âµm'], legend = True)\n",
    "    '''\n",
    "    from scipy.spatial import ConvexHull\n",
    "    # Get reads information\n",
    "    reads = pd.read_csv(readspath, index_col=0)\n",
    "    # Get a list of cell barcodes\n",
    "    cell_ids = adata.obs.index.values\n",
    "    cells_unique = np.unique(cell_ids)\n",
    "    # print(cell_ids)\n",
    "    Nlabels = cells_unique.shape[0]\n",
    "    hulls = []\n",
    "    barcodes = []\n",
    "    print('Creating cell convex hulls')\n",
    "    for i in range(len(cells_unique)):\n",
    "        cell_barcode = cells_unique[i]\n",
    "        n_reads = adata.obs.loc[cell_barcode, 'nCount_RNA']\n",
    "        if isinstance(n_reads, pd.Series):\n",
    "            n_reads = n_reads[0]\n",
    "            print(f'{cell_barcode} has multiple reads, using the first one')\n",
    "        if n_reads >= reads_min and n_reads <= reads_max:\n",
    "            barcode_num = int(cell_barcode.split('-')[1])\n",
    "            sub_reads = reads.loc[reads['cell_barcode']==barcode_num]\n",
    "            # print(sub_reads)\n",
    "            curr_coords = np.array(sub_reads[['column', 'row']]) * scale_t\n",
    "            barcodes.append(cell_barcode)\n",
    "            hulls.append(ConvexHull(curr_coords))\n",
    "    print(\"Used %d / %d\" % (len(barcodes), Nlabels))\n",
    "    return hulls, barcodes\n",
    "\n",
    "def plotHulls(adata, hulls, barcodes, color_by, color_dict,linewidth_t = 0.1,figsize_scale = 0.01, legend=None, output_path_t = None,scale_bar_t = None):\n",
    "    \"\"\" Plotting function for hulls\n",
    "    @param adata: full or subsetted adata (will take longer for more hulls)\n",
    "    @param hulls: pre-computed scipy.spatial.qhull for each unique cell (barcode)\n",
    "    @param barcodes: accompanying list of cell barcodes to relate hulls to obs in adata\n",
    "    @param color_by: name of column in adata.obs (discrete)\n",
    "    @param color_dict: dictionary of colors to use for each unique value in color_by\n",
    "    @param linewidth_t: width of hull lines\n",
    "    @param figsize_scale: scale of figure size [max(column)-min(column),max(row)-min(row)] * figsize_scale\n",
    "    @param legend: whether to include a legend\n",
    "    @param output_path_t: path to save figure\n",
    "    @param scale_bar_t: tuple of (size, units) for scale bar\n",
    "    e.g.\n",
    "        hulls, barcodes = createConvexHulls(sdata_t,reads_path,scale_t = scale_dict['63'])\n",
    "        sdata_t2 = sdata_t[np.isin(sdata_t.obs.index.values,barcodes),:].copy()\n",
    "        plotHulls(sdata_t2, hulls, barcodes, col_t, ctype_color_dict, figsize_scale = 1/200,output_path_t=f'{output_path_t}/{sample_t}_{tech_t}_convex_hull.pdf',\n",
    "        scale_bar_t=[200,'200 Âµm'], legend = True)\n",
    "    \"\"\"\n",
    "    from matplotlib.collections import PatchCollection\n",
    "    from matplotlib.patches import Patch\n",
    "    # Initialize figure\n",
    "    plt.rcParams['axes.facecolor'] = 'white'\n",
    "    fig = plt.figure(figsize=((adata.obs['column'].max()-adata.obs['column'].min())*figsize_scale,\n",
    "                        (adata.obs['row'].max() - adata.obs['row'].min())*figsize_scale))\n",
    "    ax = fig.add_subplot(111)\n",
    "    plt.xlim(adata.obs['column'].min(), adata.obs['column'].max())\n",
    "    plt.ylim(adata.obs['row'].min(), adata.obs['row'].max())\n",
    "    # Subset sample hulls accordingly \n",
    "    # Barcodes and hulls are matched, so grab index for each barcode in the subdata to get the hulls\n",
    "    idxs = [barcodes.index(i) for i in adata.obs.index]\n",
    "    plotted_hulls = [hulls[i] for i in idxs]\n",
    "    cellcolors = adata.obs[color_by]\n",
    "    polys = [hull_to_polygon_color(h, color_t = color_dict[cellcolors[i]]) for i,h in enumerate(plotted_hulls)]\n",
    "    p = PatchCollection(polys, alpha=1, linewidth=linewidth_t, match_original=True, rasterized=True)\n",
    "    ax.add_collection(p)\n",
    "    if legend:\n",
    "        clist = list(set(cellcolors))\n",
    "        clist = sorted(clist, key=lambda x:list(color_dict.keys()).index(x)) # Reorder legend\n",
    "        handles = []\n",
    "        for i in clist:\n",
    "            handles.append(Patch(color=color_dict[i], label=i))\n",
    "        ax.legend(handles=handles, title=color_by, loc='upper left', bbox_to_anchor=(1.05, 0), borderaxespad=0.)\n",
    "    plt.axis('off')\n",
    "    plt.grid(False)\n",
    "    if scale_bar_t is not None:\n",
    "        add_scale_bar(scale_bar_t[0],scale_bar_t[1])\n",
    "    if output_path_t is not None:\n",
    "        plt.savefig(f'{output_path_t}', dpi=300, bbox_inches='tight')\n",
    "    else:\n",
    "        plt.show()\n",
    "def hull_to_polygon_color(hull,k =0.9,color_t = None):\n",
    "    from matplotlib.patches import Polygon\n",
    "    cent = np.mean(hull.points, 0)\n",
    "    pts = []\n",
    "    for pt in hull.points[hull.simplices]:\n",
    "        pts.append(pt[0].tolist())\n",
    "        pts.append(pt[1].tolist())\n",
    "    pts.sort(key=lambda p: np.arctan2(p[1] - cent[1],\n",
    "                                    p[0] - cent[0]))\n",
    "    pts = pts[0::2]  # Deleting duplicates\n",
    "    pts.insert(len(pts), pts[0])\n",
    "    if color_t is not None:\n",
    "        poly = Polygon(k*(np.array(pts)- cent) + cent,edgecolor='k', linewidth=1,facecolor = color_t)\n",
    "    else:\n",
    "        poly = Polygon(k*(np.array(pts)- cent) + cent,edgecolor='k', linewidth=1)\n",
    "    poly.set_capstyle('round')\n",
    "    return poly\n",
    "def add_scale_bar(length_t,label_t,ax_t = None):\n",
    "    '''\n",
    "    Add scale bar to the plot\n",
    "    If there are multiple axes, then ax_t should be specified.\n",
    "    If ax_t is None, then the scale bar will be added to the current axis (entire plot).\n",
    "    '''\n",
    "    import matplotlib.font_manager as fm\n",
    "    from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "    fontprops = fm.FontProperties(size=20, family='Arial')\n",
    "    if ax_t is None:\n",
    "        bar = AnchoredSizeBar(plt.gca().transData, length_t, label_t, 4, pad=0.1,\n",
    "                    sep=5, borderpad=0.5, frameon=False,\n",
    "                    size_vertical=0.1, color='black',fontproperties = fontprops)\n",
    "        plt.gca().add_artist(bar)\n",
    "    else:\n",
    "        bar = AnchoredSizeBar(ax_t.transData, length_t, label_t, 4, pad=0.1,\n",
    "                    sep=5, borderpad=0.5, frameon=False,\n",
    "                    size_vertical=0.1, color='black',fontproperties = fontprops)\n",
    "        ax_t.add_artist(bar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1bf1320-ab9e-4a5b-b3fa-c46d9918cfac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change directory to the folder/region of interest \n",
    "reads_path = '/Volumes/T9/TranscriptomicsData/STARmap/36hpi_250genes_sum/lung_250genes_processed_sum/Reads_location/J/remain_reads_1.csv'\n",
    "hulls, barcodes = createConvexHulls(adata,reads_path)\n",
    "adata_sub = adata[barcodes,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c432d6e-4fc1-4262-81b6-4e3ccb6bd72b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot convex hull representing cell types \n",
    "cell_color = {\n",
    "               \"AirwayEpithelial\": \"#DB5F57\",\n",
    "               \"AT1\" :\"#db8557\",\n",
    "               \"AT2\":\"#dbaa57\",\n",
    "               \"BronchoFibroblast\":\"#dbd057\",\n",
    "               \"AlveolarFibroblast\":\"#c0db57\",\n",
    "               \"MyoFB\":\"#9bdb57\",\n",
    "               \"Pericyte\":\"#75db57\",\n",
    "               \"SMC\":\"#57db5f\",\n",
    "               \"Artery\":\"#57db85\",\n",
    "               \"Vein\":\"#57dbaa\",\n",
    "               \"Capillary\":\"#57dbd0\",\n",
    "               \"AlveolarMac\":\"#57c0db\",\n",
    "               \"InterstitialMac\":\"#579bdb\",\n",
    "               \"InflamMac\":\"#5775db\",\n",
    "               \"PatrolMac\":\"#5f57db\",\n",
    "               \"mucosalDC\":\"#8557db\",\n",
    "               \"perivascularDC\":\"#aa57db\",\n",
    "               \"pDC\":\"#d057db\",\n",
    "               \"Neutrophil\":\"#db57c0\",\n",
    "               \"Tcell\":\"#db579b\",\n",
    "               \"Bcell\":\"#db5775\", \n",
    "}\n",
    "\n",
    "plotHulls(adata_sub, hulls, barcodes, 'celltype', cell_color, legend = True, figsize_scale = 1/200,scale_bar_t=[200,'200 Âµm'])#, output_path_t=f'2500210_regionD_celltype.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0aceb073-f816-4c37-961a-cced1cf6dc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot convex hull representing normlaized flu level\n",
    "color = sns.color_palette(\"rocket_r\", 10)\n",
    "cell_color = {\n",
    "    1:color[0], \n",
    "    2:color[1], \n",
    "    3:color[2], \n",
    "    4:color[3], \n",
    "    5:color[4],\n",
    "    6:color[5], \n",
    "    7:color[6], \n",
    "    8:color[7], \n",
    "    9:color[8], \n",
    "    10:color[9]           \n",
    "}\n",
    "plotHulls(adata_sub, hulls, barcodes, 'flu', cell_color, legend = True, figsize_scale = 1/200,scale_bar_t=[200,'200 Âµm'])#,output_path_t=f'regionA_np_prot.pdf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "598f3f81-a523-4480-95db-a33f0e45beb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot convex hull representing normalized ifn level\n",
    "color = sns.color_palette(\"rocket_r\", 10)\n",
    "cell_color = {\n",
    "    1:color[0], \n",
    "    2:color[1], \n",
    "    3:color[2], \n",
    "    4:color[3], \n",
    "    5:color[4],\n",
    "    6:color[5], \n",
    "    7:color[6], \n",
    "    8:color[7], \n",
    "    9:color[8], \n",
    "    10:color[9]           \n",
    "}\n",
    "plotHulls(adata_sub, hulls, barcodes, 'ifn', cell_color, legend = True, figsize_scale = 1/200,scale_bar_t=[200,'200 Âµm'])#,output_path_t=f'regionA_np_prot.pdf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa5ff600-2acc-4762-ad89-31f8a78f97a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find clusters of adjacent infected cells using DBSCAN\n",
    "from sklearn.cluster import DBSCAN\n",
    "import numpy as np\n",
    "\n",
    "ind = flu_exp > 1.2\n",
    "infect = cop.loc[ind]\n",
    "infect_cell = celltype.loc[ind]\n",
    "infect_meta = meta.loc[ind]\n",
    "sns.scatterplot(x=infect_meta['x'],y=infect_meta['y'],hue=infect['NP.gRNA'], size =1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d88d17-ce17-4714-bb66-a2ee239ecd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "infect_coord = pd.concat([infect_meta['x'],infect_meta['y']], axis=1)\n",
    "coord = np.array(infect_coord)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad1be21a-6ef9-4f0e-bd86-08a55c0fc977",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cells within 175 pixel of distance (20 um) are considered neighbors; \n",
    "# clusters of at least 3 cells are considered \n",
    "\n",
    "dbscan = DBSCAN(eps=175, min_samples=3) \n",
    "clusters = dbscan.fit_predict(coord)\n",
    "\n",
    "clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e903ea6c-2ae9-4e5d-be50-820bf36b5b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(x=infect_meta['x'],y=infect_meta['y'],hue=clusters, size =5)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f69bac43-bd34-4271-8d7a-ed59610cf0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "clust = np.unique(clusters)\n",
    "clust = clust[1:len(clust)]\n",
    "clust"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "766485d3-67f3-43f0-b25a-d82659231859",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the composition of infected cell types for each plaque \n",
    "blank = np.zeros((len(clust), 21))\n",
    "plaque_info = pd.DataFrame(blank, columns = [\"AirwayEpithelial\",\n",
    "               \"AT1\",\n",
    "               \"AT2\",\n",
    "               \"BronchoFibroblast\",\n",
    "               \"AlveolarFibroblast\",\n",
    "               \"MyoFB\",\n",
    "               \"Pericyte\",\n",
    "               \"SMC\",\n",
    "               \"Artery\",\n",
    "               \"Vein\",\n",
    "               \"Capillary\",\n",
    "               \"AlveolarMac\",\n",
    "               \"InterstitialMac\",\n",
    "               \"InflamMac\",\n",
    "               \"PatrolMac\",\n",
    "               \"mucosalDC\",\n",
    "               \"perivascularDC\",\n",
    "               \"pDC\",\n",
    "               \"Neutrophil\",\n",
    "               \"Tcell\",\n",
    "               \"Bcell\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75e4da8-693a-4871-ac19-19f8c96c797d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate through each cluster and count the number of infected cells for each cell type \n",
    "for i in clust:\n",
    "    sub_ind = clusters == i\n",
    "    sub_infect = infect.loc[sub_ind]\n",
    "    sub_cell = infect_cell.loc[sub_ind]\n",
    "    cell_infect = sub_cell['x'].unique()\n",
    "    num_infect = sub_cell['x'].value_counts(sort=False)\n",
    "    for j in range(len(cell_infect)):\n",
    "        plaque_info.loc[i, cell_infect[j]] = num_infect.iloc[j]\n",
    "plaque_info.to_csv(region+'_plaque_info.csv')\n",
    "plaque_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6af59206-e710-4786-85a7-6aaf24731b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the composition of IFN+ cell types for each plaque \n",
    "blank = np.zeros((len(clust), 21))\n",
    "ifn_info = pd.DataFrame(blank, columns = [\"AirwayEpithelial\",\n",
    "               \"AT1\",\n",
    "               \"AT2\",\n",
    "               \"BronchoFibroblast\",\n",
    "               \"AlveolarFibroblast\",\n",
    "               \"MyoFB\",\n",
    "               \"Pericyte\",\n",
    "               \"SMC\",\n",
    "               \"Artery\",\n",
    "               \"Vein\",\n",
    "               \"Capillary\",\n",
    "               \"AlveolarMac\",\n",
    "               \"InterstitialMac\",\n",
    "               \"InflamMac\",\n",
    "               \"PatrolMac\",\n",
    "               \"mucosalDC\",\n",
    "               \"perivascularDC\",\n",
    "               \"pDC\",\n",
    "               \"Neutrophil\",\n",
    "               \"Tcell\",\n",
    "               \"Bcell\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "743b2aa8-6c44-4969-bdca-78a54ba57023",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in clust:\n",
    "    sub_ind = clusters == i\n",
    "    sub_infect = infect.loc[sub_ind]\n",
    "    sub_cell = infect_cell.loc[sub_ind]\n",
    "    ifn_pos = sub_infect[IFN].sum(axis=1) > 0\n",
    "    sub_ifn_cell = sub_cell.loc[ifn_pos]\n",
    "    num_ifn = sub_ifn_cell['x'].value_counts(sort=False)\n",
    "    cell_ifn = sub_ifn_cell['x'].unique()\n",
    "    for j in range(len(cell_ifn)):\n",
    "        ifn_info.loc[i, cell_ifn[j]] = num_ifn.iloc[j]\n",
    "ifn_info.T.to_csv(region+'_ifn_info.csv')\n",
    "ifn_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e73c5a-10b7-4006-8db4-3b7a856a0b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_ifn = ifn_info / plaque_info * 100\n",
    "frac_ifn.to_csv(region+'_frac_ifn.csv')\n",
    "frac_ifn"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
